Build with ghc(js/wasm) 9.8.2 + 9.10.1 + 9.12.2 + 9.14.1

ymeister opened the pull request:
Needs:
- https://github.com/reflex-frp/reflex/pull/502
- https://github.com/reflex-frp/patch/pull/56 (or https://github.com/reflex-frp/patch/pull/55).
_____

For `haskell.nix`:
`default.nix`:
```
let deps = {
      "haskell.nix" = builtins.fetchTarball {
        url = "https://github.com/input-output-hk/haskell.nix/archive/d8c50dcaf3d3d589829ee9be9d5dba8279b8cc59.tar.gz";
        sha256 = "0a5hgryz6nszmy67yf1aks399h2aw0nj845518c4prs5c6ns1z7p";
      };
      patch = pkgs.fetchFromGitHub {
        owner = "ymeister";
        repo = "patch";
        rev = "8ebd54a5fca390518b9ac4e764b88034990fa277";
        sha256 = "16wr2x9zg0dshq3wq94bm7hjj94ld030av011zkikand6ri4jd5f";
      };
      reflex = pkgs.fetchFromGitHub {
        owner = "ymeister";
        repo = "reflex";
        rev = "315bef0c18e04a0491dfb496bc5163b0cf49ab17";
        sha256 = "1glc5g67gw7ksc22ilyl7xjzw12s59b9di39j19pg5jl1hkkc8z2";
      };
    };

    haskellNix = import deps."haskell.nix" {};

    # Import nixpkgs and pass the haskell.nix provided nixpkgsArgs
    pkgs = import
      # haskell.nix provides access to the nixpkgs pins which are used by our CI,
      # hence you will be more likely to get cache hits when using these.
      # But you can also just use your own, e.g. '<nixpkgs>'.
      haskellNix.sources.nixpkgs-unstable
      # These arguments passed to nixpkgs, include some patches and also
      # the haskell.nix functionality itself as an overlay.
      haskellNix.nixpkgsArgs;

    source-repository-packages = packages:
      builtins.zipAttrsWith
        (k: vs:
          if k == "cabalProjectLocal" then pkgs.lib.strings.concatStringsSep "\n" vs
          else builtins.zipAttrsWith (_: pkgs.lib.lists.last) vs
        )
        (pkgs.lib.lists.forEach packages (p:
          let input = builtins.unsafeDiscardStringContext p;
          in {
            inputMap."${input}" = { name = builtins.baseNameOf p; outPath = p; rev = "HEAD"; };
            cabalProjectLocal = ''
              source-repository-package
                type: git
                location: ${input}
                tag: HEAD
            '';
          }
        ));

    import-cabal-project = dir: file:
      let path = dir + "/${file}";
          content = ''
            -- ${path}
            ${builtins.readFile path}
          '';
          lines = pkgs.lib.strings.splitString "\n" content;
      in pkgs.lib.strings.concatStringsSep "\n" (
          pkgs.lib.lists.forEach lines (line:
            if pkgs.lib.strings.hasPrefix "import: " line
            then import-cabal-project dir (pkgs.lib.strings.removePrefix "import: " line)
            else line
          )
      );

    haskellDeps = source-repository-packages [
      deps.reflex
      deps.patch
    ];

    project = pkgs: pkgs.haskell-nix.project {
      src = ./.;

      inherit (haskellDeps) inputMap cabalProjectLocal;
      cabalProject = import-cabal-project ./. "cabal.project";

      shell.withHaddock = if pkgs.stdenv.hostPlatform.isGhcjs then false else true;

      modules = [({ pkgs, lib, ... }: {
        packages.reflex-dom-core.components.tests = {
          gc.buildable = lib.mkForce false;
          hydration.buildable = lib.mkForce false;
          #gc.build-tools = [ pkgs.chromium ];
          #hydration.build-tools = [ pkgs.chromium ];
        };
      })];

      compiler-nix-name = "ghc910";
    };

in {
  ghc = project pkgs;
  ghc-js = project pkgs.pkgsCross.ghcjs;
}
```
`shell.nix`:
```
let project = import ./default.nix;
in project.ghc.shellFor {
  tools = {
    cabal = "latest";
  };

  buildInputs = with project.pkgs; [
    nodejs-slim_latest
  ];

  # Sellect cross compilers to include.
  crossPlatforms = ps: with ps; [
    ghcjs # Adds support for `javascript-unknown-ghcjs-cabal build` in the shell
  ];
}
```
_____

For `nixpkgs`:

From `patch` PR:
>When building on ghc js, requires single-core building (https://gitlab.haskell.org/ghc/ghc/-/issues/25083#note_578275).
>
>For nix:
>```
>patch = haskellLib.overrideCabal (drv: { enableParallelBuilding = false; }) super.patch;
>```

Fix `os-string` (https://github.com/NixOS/nixpkgs/pull/332123):
```
os-string = null;
```
_____

To build a project depending on `reflex-dom` using `haskell.nix`:
`default.nix`:
```
let deps = {
      "haskell.nix" = builtins.fetchTarball {
        url = "https://github.com/input-output-hk/haskell.nix/archive/d8c50dcaf3d3d589829ee9be9d5dba8279b8cc59.tar.gz";
        sha256 = "0a5hgryz6nszmy67yf1aks399h2aw0nj845518c4prs5c6ns1z7p";
      };
      patch = pkgs.fetchFromGitHub {
        owner = "ymeister";
        repo = "patch";
        rev = "8ebd54a5fca390518b9ac4e764b88034990fa277";
        sha256 = "16wr2x9zg0dshq3wq94bm7hjj94ld030av011zkikand6ri4jd5f";
      };
      reflex = pkgs.fetchFromGitHub {
        owner = "ymeister";
        repo = "reflex";
        rev = "315bef0c18e04a0491dfb496bc5163b0cf49ab17";
        sha256 = "1glc5g67gw7ksc22ilyl7xjzw12s59b9di39j19pg5jl1hkkc8z2";
      };
      reflex-dom = pkgs.fetchFromGitHub {
        owner = "ymeister";
        repo = "reflex-dom";
        rev = "b69b759a0d90d3ac7fc35083db985603088afda1";
        sha256 = "092vqav1pnsnsrxawxb88pak4sa9igr5033hg95bypl3snqx1246";
      };
    };

    haskellNix = import deps."haskell.nix" {};

    # Import nixpkgs and pass the haskell.nix provided nixpkgsArgs
    pkgs = import
      # haskell.nix provides access to the nixpkgs pins which are used by our CI,
      # hence you will be more likely to get cache hits when using these.
      # But you can also just use your own, e.g. '<nixpkgs>'.
      haskellNix.sources.nixpkgs-unstable
      # These arguments passed to nixpkgs, include some patches and also
      # the haskell.nix functionality itself as an overlay.
      haskellNix.nixpkgsArgs;

    source-repository-packages = packages:
      builtins.zipAttrsWith
        (k: vs:
          if k == "cabalProjectLocal" then pkgs.lib.strings.concatStringsSep "\n" vs
          else builtins.zipAttrsWith (_: pkgs.lib.lists.last) vs
        )
        (pkgs.lib.lists.forEach packages (p:
          let input = builtins.unsafeDiscardStringContext p;
          in {
            inputMap."${input}" = { name = builtins.baseNameOf p; outPath = p; rev = "HEAD"; };
            cabalProjectLocal = ''
              source-repository-package
                type: git
                location: ${input}
                tag: HEAD
            '';
          }
        ));

    import-cabal-project = dir: file:
      let path = dir + "/${file}";
          content = ''
            -- ${path}
            ${builtins.readFile path}
          '';
          lines = pkgs.lib.strings.splitString "\n" content;
      in pkgs.lib.strings.concatStringsSep "\n" (
          pkgs.lib.lists.forEach lines (line:
            if pkgs.lib.strings.hasPrefix "import: " line
            then import-cabal-project dir (pkgs.lib.strings.removePrefix "import: " line)
            else line
          )
      );

    haskellDeps = source-repository-packages [
      (deps.reflex-dom + "/reflex-dom")
      (deps.reflex-dom + "/reflex-dom-core")
      deps.reflex
      deps.patch
    ];

    project = pkgs: pkgs.haskell-nix.project {
      src = ./.;

      inherit (haskellDeps) inputMap;
      cabalProject = import-cabal-project ./. "cabal.project";
      cabalProjectLocal = ''
        ${import-cabal-project deps.reflex-dom "cabal.dependencies.project"}

        ${haskellDeps.cabalProjectLocal}
      '';

      shell.withHaddock = if pkgs.stdenv.hostPlatform.isGhcjs then false else true;

      modules = [({ pkgs, lib, ... }: {
        packages.reflex-dom-core.components.tests = {
          gc.buildable = lib.mkForce false;
          hydration.buildable = lib.mkForce false;
          #gc.build-tools = [ pkgs.chromium ];
          #hydration.build-tools = [ pkgs.chromium ];
        };
      })];

      compiler-nix-name = "ghc910";
    };

in {
  ghc = project pkgs;
  ghc-js = project pkgs.pkgsCross.ghcjs;
}
```
`shell.nix`:
```
let project = import ./default.nix;
in project.ghc.shellFor {
  tools = {
    cabal = "latest";
  };

  buildInputs = with project.pkgs; [
    nodejs-slim_latest
  ];

  # Sellect cross compilers to include.
  crossPlatforms = ps: with ps; [
    ghcjs # Adds support for `javascript-unknown-ghcjs-cabal build` in the shell
  ];
}
```

ymeister committed:
Remove JavaScriptFFI pragma

ymeister committed:
Make reflex-dom buildable on ghc js backend

ymeister committed:
Use Type from Data.Kind instead of *

ymeister committed:
Cleanup warnings

ymeister committed:
Build with ghc(js)-9.10

ymeister committed:
Update bounds with 'cabal-bounds' & recover backwards compatibility

ymeister commented:
Needs https://github.com/reflex-frp/reflex/pull/502 to be in hackage to pass the CI.

odr commented:
How to check it? 
I took `default.nix` and `shell.nix` from the end of PR description. Then I run `cabal init` and add `reflex-dom` into cabal deps. Also add `cabal.project`.
Next, I open nix-shell (successfully, took 2h), run `javascript-unknown-ghcjs-cabal build` and get (similar to [this](https://gitlab.haskell.org/ghc/ghc/-/issues/25083)):
```
src/Data/Patch/IntMap.hs:1:1: error: [GHC-87897]
    Exception when trying to run compile-time code:
      getBin: Unknown encoding for constructor
    Code: makeWrapped ''PatchIntMap
```
Is there a way to build now?

ymeister commented:
@odr 
> How to check it? I took `default.nix` and `shell.nix` from the end of PR description. Then I run `cabal init` and add `reflex-dom` into cabal deps. Also add `cabal.project`. Next, I open nix-shell (successfully, took 2h), run `javascript-unknown-ghcjs-cabal build` and get (similar to [this](https://gitlab.haskell.org/ghc/ghc/-/issues/25083)):
> 
> ```
> src/Data/Patch/IntMap.hs:1:1: error: [GHC-87897]
>     Exception when trying to run compile-time code:
>       getBin: Unknown encoding for constructor
>     Code: makeWrapped ''PatchIntMap
> ```
> 
> Is there a way to build now?

~~Hmm, `nix-shell` being successful means that `patch` should have already been compiled successfully, both for ghc and ghc-js.~~ If it's the same problem as in https://gitlab.haskell.org/ghc/ghc/-/issues/25083 , then compiling with `javascript-unknown-ghcjs-cabal build -j1` should resolve the issue (although I doubt it's the case). The fastest and most reliable way for me to figure out what's wrong with your case would be if you could give me some stripped down version of the project you're trying to build, be it via repo, branch or sending me an archive.

P.S. Make sure you're using the **latest** `default.nix`/`shell.nix` from PR description. There were some edits done to it.

odr commented:
I'll share code little bit later. Now I just tried `-j1` and add `-v3`. Now an error: 
```
...
[ 4 of 12] Compiling Data.Patch.IntMap ( src/Data/Patch/IntMap.hs, dist/build/Data/Patch/IntMap.o )
ghc-interp.js: Message not supported with the JavaScript interpreter: InitLinker
CallStack (from HasCallStack):
  error, called at libraries/ghci/GHCi/Run.hs:142:20 in ghci-9.10.1-inplace:GHCi.Run
HasCallStack backtrace:
  collectBacktraces, called at libraries/ghc-internal/src/GHC/Internal/Exception.hs:92:13 in ghc-internal:GHC.Internal.Exception
  toExceptionWithBacktrace, called at libraries/ghc-internal/src/GHC/Internal/Exception.hs:128:3 in ghc-internal:GHC.Internal.Exception

src/Data/Patch/IntMap.hs:1:1: error: [GHC-87897]
    Exception when trying to run compile-time code:
      External interpreter terminated (1)
    Code: makeWrapped ''PatchIntMap
  |
1 | {-# LANGUAGE CPP #-}
  | ^
...
```

odr commented:
@ymeister My project: https://github.com/odr/reflex-js. It is empty. I just added `default.nix` and `shell.nix`,  then create `.cabal` and `cabal.project` and add `reflex-dom` into dependencies.
I still get an errors when build:
```
[nix-shell:~/git/reflex-js]$ javascript-unknown-ghcjs-cabal build    
Build profile: -w ghc-9.10.1 -O1
In order, the following will be built (use -v for more details):
 - patch-0.0.8.2 (lib) (requires build)

...

[ 4 of 12] Compiling Data.Patch.IntMap ( src/Data/Patch/IntMap.hs, dist/build/Data/Patch/IntMap.o )
src/Data/Patch/IntMap.hs:1:1: error: [GHC-87897]
    Exception when trying to run compile-time code:
      {handle: <file descriptor: 35>}: GHCi.Message.remoteCall: end of file
    Code: makeWrapped ''PatchIntMap
  |
1 | {-# LANGUAGE CPP #-}
  | ^

...
Error: [Cabal-7125]
Failed to build patch-0.0.8.2 (which is required by exe:reflex-js from reflex-js-0.1.0.0). See the build log above for details.



[nix-shell:~/git/reflex-js]$ javascript-unknown-ghcjs-cabal build -j1
Build profile: -w ghc-9.10.1 -O1
In order, the following will be built (use -v for more details):
 - patch-0.0.8.2 (lib) (requires build)

...

[ 4 of 12] Compiling Data.Patch.IntMap ( src/Data/Patch/IntMap.hs, dist/build/Data/Patch/IntMap.o )
ghc-interp.js: Message not supported with the JavaScript interpreter: InitLinker
CallStack (from HasCallStack):
  error, called at libraries/ghci/GHCi/Run.hs:142:20 in ghci-9.10.1-inplace:GHCi.Run
HasCallStack backtrace:
  collectBacktraces, called at libraries/ghc-internal/src/GHC/Internal/Exception.hs:92:13 in ghc-internal:GHC.Internal.Exception
  toExceptionWithBacktrace, called at libraries/ghc-internal/src/GHC/Internal/Exception.hs:128:3 in ghc-internal:GHC.Internal.Exception

src/Data/Patch/IntMap.hs:1:1: error: [GHC-87897]
    Exception when trying to run compile-time code:
      External interpreter terminated (1)
    Code: makeWrapped ''PatchIntMap
  |
1 | {-# LANGUAGE CPP #-}
  | ^

...

Error: [Cabal-7125]
Failed to build patch-0.0.8.2 (which is required by exe:reflex-js from reflex-js-0.1.0.0).


```

ymeister commented:
@odr Thank you for providing an example! That said, I did not encounter any problem building it with neither native ghc nor ghc-js. Could you please try deleting `dist` and/or `dist-newstyle` directories generated by `cabal build`, and try to build it again?

odr commented:
> @odr Thank you for providing an example! That said, I did not encounter any problem building it with neither native ghc nor ghc-js. Could you please try deleting `dist` and/or `dist-newstyle` directories generated by `cabal build`, and try to build it again?

I removed `dist-newstyle` and repeat with `-j1`. The same...

How do you build literally? From nix-shell?

I'll ask my colleague to try this tomorrow...

ymeister commented:
@odr 

I literally just:
1. `git clone https://github.com/odr/reflex-js.git`
2. `cd reflex-js`
3. `nix-shell -v`
4. `javascript-unknown-ghcjs-cabal build`

That's all. My system is running on `nixos-unstable` (`rev`: `cb9a96f23c491c081b38eab96d22fa958043c9fa`), if that is of any help. Although, the whole thing about `nix` is reproducibility and all the project dependencies have been pinned, so I don't know what's interfering with your side. Your error messages say something about errors with JS interpreter, maybe you have `nodejs` installed on your system that somehow interferes with something? Although the `nodejs_slim_latest` from `shell.nix` should have overrode this.

odr commented:
@ymeister 
I checked my node.js and don't see any problem:
```
[nix-shell:~/git/reflex-js]$ which node
/nix/store/9ficrfvgnm4q91wgw3bmpfd34dyg6bbz-nodejs-slim-22.3.0/bin/node
[nix-shell:~/git/reflex-js]$ which nodejs
/usr/local/bin/nodejs
[nix-shell:~/git/reflex-js]$ ls -la /usr/local/bin/nodejs 
lrwxrwxrwx 1 root root 19 Aug 15 12:20 /usr/local/bin/nodejs -> /usr/local/bin/node
[nix-shell:~/git/reflex-js]$ node --version
v22.3.0
[nix-shell:~/git/reflex-js]$ nodejs --version
v22.6.0
```
FI: I am on Ubuntu-22.04. AMD Ryzen 7 4800H, cores=8, threads=16

I also tried to run `nix-shell --pure` (with adding `git which nix` into `buildInputs`) but result is the same.

odr commented:
@ymeister I built it without problems on the new laptop! There is Ubuntu 24.04,  no nodejs installed, multi-user nix setup (on the old one I had single-user setup).
So I am unblocked now but I heard about problems with build from others (e.g. on MacOS)...

ali-abrar committed:
ci: Exclude some macos + old ghc combos

ali-abrar commented:
@ymeister could you resolve conflicts?

ymeister commented:
> @ymeister could you resolve conflicts?

Marked it in my todo list.

ymeister committed:
Add missing pragmas

ymeister committed:
Fixup pragma

ymeister committed:
Update bounds to support ghc 9.14

ymeister committed:
Fix hydration tests to run without network namespace isolation

- Add -threaded to hydration test ghc-options
- Use dynamic port allocation instead of fixed jsaddlePort to avoid
  port conflicts and browser caching between tests
- Remove unshareNetwork dependency so tests can run without Linux
  namespace support
- Re-enable "places fences and removes them" test
- Add upper bound on webdriver dependency

ymeister committed:
Fixup pragmas

ymeister committed:
Don't build hlint test on wasm

ymeister committed:
GHC 9.14 + WASM

ymeister committed:
Use `nix-haskell` driver

ymeister committed:
Update Github CI

ymeister committed:
Add flake inputs and shell.nix for nix-haskell integration

- Define explicit flake inputs following nix-haskell conventions
- Add inputs.nix for flake-compat bootstrapping outside of flakes
- Add shell.nix for convenient `nix-shell` entry point
- Pass `inputs` through default.nix to nix-haskell driver
- Update nix-haskell submodule

ymeister committed:
Update `nix-haskell` submodule and make it use `http(s)` url instead of `ssh`

ymeister committed:
Update `nix-haskell` pin and fixup flake submodules
